jvm/src/main/scala-2.10/scalajson.ast/JValue.scala

```diff
@@ -0,0 +1,211 @@
+package scalajson.ast
+
+/** Represents a valid JSON Value
+  *
+  * @author Matthew de Detrich
+  * @see https://www.ietf.org/rfc/rfc4627.txt
+  */
+sealed abstract class JValue extends Product with Serializable {
+
+  /**
+    * Converts a [[JValue]] to a [[unsafe.JValue]]. Note that
+    * when converting [[JObject]], this can produce [[unsafe.JObject]] of
+    * unknown ordering, since ordering on a [[scala.collection.Map]] isn't defined. Duplicate keys will also
+    * be removed in an undefined manner.
+    *
+    * @see https://www.ietf.org/rfc/rfc4627.txt
+    * @return
+    */
+  def toUnsafe: unsafe.JValue
+}
+
+/** Represents a JSON null value
+  *
+  * @author Matthew de Detrich
+  */
+final case object JNull extends JValue {
+  override def toUnsafe: unsafe.JValue = unsafe.JNull
+}
+
+/** Represents a JSON string value
+  *
+  * @author Matthew de Detrich
+  */
+final case class JString(value: String) extends JValue {
+  override def toUnsafe: unsafe.JValue = unsafe.JString(value)
+}
+
+/**
+  * If you are passing in a NaN or Infinity as a Double, JNumber will
+  * return a JNull
+  */
+object JNumber {
+  def apply(value: Int): JNumber = new JNumber(value.toString)
+
+  def apply(value: Long): JNumber = new JNumber(value.toString)
+
+  def apply(value: BigInt): JNumber = new JNumber(value.toString)
+
+  /**
+    * @param value
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
+    */
+  def apply(value: Float): JValue = value match {
+    case n if java.lang.Float.isNaN(n) => JNull
+    case n if n.isInfinity => JNull
+    case _ => new JNumber(value.toString)
+  }
+
+  def apply(value: BigDecimal): JNumber = new JNumber(value.toString())
+
+  /**
+    * @param value
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
+    */
+  def apply(value: Double): JValue = value match {
+    case n if n.isNaN => JNull
+    case n if n.isInfinity => JNull
+    case _ => new JNumber(value.toString)
+  }
+
+  def apply(value: Integer): JNumber = new JNumber(value.toString)
+
+  def apply(value: Array[Char]): Option[JNumber] = {
+    val s = new String(value)
+    if (s.matches(jNumberRegex))
+      Some(new JNumber(s))
+    else
+      None
+  }
+
+  def fromString(value: String): Option[JNumber] =
+    if (value.matches(jNumberRegex))
+      Some(new JNumber(value))
+    else
+      None
+
+  def unapply(arg: JNumber): Option[String] = Some(arg.value)
+}
+
+/** Represents a JSON number value. If you are passing in a
+  * NaN or Infinity as a [[scala.Double]] or [[scala.Float]], [[JNumber]] will
+  * return a [[JNull]].
+  *
+  * @author Matthew de Detrich
+  */
+// Due to a restriction in Scala 2.10, we can't use a proper case class along with
+// a private constructor, so we have to simulate a case class with a normal class
```

Comment:
As a side note and for the record: This is not correct. Even in Scala 2.10 case classes can have a private constructor. The issue is that we cannot override/replace the default `apply` method generated by the compiler even when the constructor itself is marked `private`, which kind of defeats the purpose of marking the constructor `private` in the first place.